ConcurrentModificationException during startup from AddonSuggestionService

I saw this in my log file after starting openHAB 4.1.0.RC1:

```
2023-12-21 07:53:42.038 [ERROR] [scovery.addon.AddonSuggestionService] - bundle org.openhab.core.config.discovery.addon:4.1.0.RC1 (167)[org.openhab.core.config.discovery.addon.AddonSuggestionService(127)] : The addAddonFinder method has thrown an exception
java.util.ConcurrentModificationException: null
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1631) ~[?:?]
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509) ~[?:?]
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150) ~[?:?]
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173) ~[?:?]
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596) ~[?:?]
        at org.openhab.core.config.discovery.addon.AddonSuggestionService.changed(AddonSuggestionService.java:189) ~[?:?]
        at org.openhab.core.config.discovery.addon.AddonSuggestionService.addAddonFinder(AddonSuggestionService.java:177) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[?:?]
        at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:568) ~[?:?]
```

--- a/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
+++ b/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
@@ -73,6 +73,7 @@ public class AddonSuggestionService implements AutoCloseable {
     private final ScheduledExecutorService scheduler;
     private final Map<String, Boolean> baseFinderConfig = new ConcurrentHashMap<>();
     private final List<Future<?>> tasks = new CopyOnWriteArrayList<>();
+    private final Object synchObject = new Object();
 
     @Activate
     public AddonSuggestionService(final @Reference ConfigurationAdmin configurationAdmin,
@@ -161,7 +162,7 @@ public class AddonSuggestionService implements AutoCloseable {
 
     @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
     public void addAddonInfoProvider(AddonInfoProvider addonInfoProvider) {
-        synchronized (this) {
+        synchronized (synchObject) {
             addonInfoProviders.add(addonInfoProvider);
         }
         changed();
@@ -169,7 +170,7 @@ public class AddonSuggestionService implements AutoCloseable {
 
     public void removeAddonInfoProvider(AddonInfoProvider addonInfoProvider) {
         boolean removed;
-        synchronized (this) {
+        synchronized (synchObject) {
             removed = addonInfoProviders.remove(addonInfoProvider);
         }
         if (removed) {
@@ -179,7 +180,7 @@ public class AddonSuggestionService implements AutoCloseable {
 
     @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
     public void addAddonFinder(AddonFinder addonFinder) {
-        synchronized (this) {
+        synchronized (synchObject) {
             addonFinders.add(addonFinder);
         }
         changed();
@@ -187,7 +188,7 @@ public class AddonSuggestionService implements AutoCloseable {
 
     public void removeAddonFinder(AddonFinder addonFinder) {
         boolean removed;
-        synchronized (this) {
+        synchronized (synchObject) {
             removed = addonFinders.remove(addonFinder);
         }
         if (removed) {
@@ -196,7 +197,7 @@ public class AddonSuggestionService implements AutoCloseable {
     }
 
     private void changed() {
-        synchronized (this) {
+        synchronized (synchObject) {
             List<AddonInfo> candidates = addonInfoProviders.stream()
                     .map(p -> p.getAddonInfos(localeProvider.getLocale())).flatMap(Collection::stream).toList();
             addonFinders.stream().filter(this::isFinderEnabled).forEach(f -> f.setAddonCandidates(candidates));
@@ -206,14 +207,14 @@ public class AddonSuggestionService implements AutoCloseable {
     @Deactivate
     @Override
     public void close() throws Exception {
-        synchronized (this) {
+        synchronized (synchObject) {
             addonFinders.clear();
             addonInfoProviders.clear();
         }
     }
 
     public Set<AddonInfo> getSuggestedAddons(@Nullable Locale locale) {
-        synchronized (this) {
+        synchronized (synchObject) {
             return addonFinders.stream().filter(this::isFinderEnabled).map(f -> f.getSuggestedAddons())
                     .flatMap(Collection::stream).collect(Collectors.toSet());
         }

--- a/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
+++ b/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
@@ -184,9 +184,11 @@ public class AddonSuggestionService implements AutoCloseable {
     }
 
     private void changed() {
-        List<AddonInfo> candidates = addonInfoProviders.stream().map(p -> p.getAddonInfos(localeProvider.getLocale()))
-                .flatMap(Collection::stream).toList();
-        addonFinders.stream().filter(this::isFinderEnabled).forEach(f -> f.setAddonCandidates(candidates));
+        synchronized (addonInfoProviders) {
+            List<AddonInfo> candidates = addonInfoProviders.stream()
+                    .map(p -> p.getAddonInfos(localeProvider.getLocale())).flatMap(Collection::stream).toList();
+            addonFinders.stream().filter(this::isFinderEnabled).forEach(f -> f.setAddonCandidates(candidates));
+        }
     }
 
     @Deactivate

--- a/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
+++ b/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
@@ -206,12 +206,16 @@ public class AddonSuggestionService implements AutoCloseable {
     @Deactivate
     @Override
     public void close() throws Exception {
-        addonFinders.clear();
-        addonInfoProviders.clear();
+        synchronized (this) {
+            addonFinders.clear();
+            addonInfoProviders.clear();
+        }
     }
 
     public Set<AddonInfo> getSuggestedAddons(@Nullable Locale locale) {
-        return addonFinders.stream().filter(this::isFinderEnabled).map(f -> f.getSuggestedAddons())
-                .flatMap(Collection::stream).collect(Collectors.toSet());
+        synchronized (this) {
+            return addonFinders.stream().filter(this::isFinderEnabled).map(f -> f.getSuggestedAddons())
+                    .flatMap(Collection::stream).collect(Collectors.toSet());
+        }
     }
 }

--- a/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
+++ b/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
@@ -180,12 +180,8 @@ public class AddonSuggestionService implements AutoCloseable {
     }
 
     public void removeAddonFinder(AddonFinder addonFinder) {
-        boolean removed;
         synchronized (addonFinders) {
-            removed = addonFinders.remove(addonFinder);
-        }
-        if (removed) {
-            changed();
+            addonFinders.remove(addonFinder);
         }
     }
 

--- a/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
+++ b/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
@@ -183,12 +183,10 @@ public class AddonSuggestionService implements AutoCloseable {
         }
     }
 
-    private void changed() {
-        synchronized (addonInfoProviders) {
-            List<AddonInfo> candidates = addonInfoProviders.stream()
-                    .map(p -> p.getAddonInfos(localeProvider.getLocale())).flatMap(Collection::stream).toList();
-            addonFinders.stream().filter(this::isFinderEnabled).forEach(f -> f.setAddonCandidates(candidates));
-        }
+    private synchronized void changed() {
+        List<AddonInfo> candidates = addonInfoProviders.stream().map(p -> p.getAddonInfos(localeProvider.getLocale()))
+                .flatMap(Collection::stream).toList();
+        addonFinders.stream().filter(this::isFinderEnabled).forEach(f -> f.setAddonCandidates(candidates));
     }
 
     @Deactivate

--- a/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
+++ b/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
@@ -161,32 +161,46 @@ public class AddonSuggestionService implements AutoCloseable {
 
     @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
     public void addAddonInfoProvider(AddonInfoProvider addonInfoProvider) {
-        addonInfoProviders.add(addonInfoProvider);
+        synchronized (this) {
+            addonInfoProviders.add(addonInfoProvider);
+        }
         changed();
     }
 
     public void removeAddonInfoProvider(AddonInfoProvider addonInfoProvider) {
-        if (addonInfoProviders.remove(addonInfoProvider)) {
+        boolean removed;
+        synchronized (this) {
+            removed = addonInfoProviders.remove(addonInfoProvider);
+        }
+        if (removed) {
             changed();
         }
     }
 
     @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
     public void addAddonFinder(AddonFinder addonFinder) {
-        addonFinders.add(addonFinder);
+        synchronized (this) {
+            addonFinders.add(addonFinder);
+        }
         changed();
     }
 
     public void removeAddonFinder(AddonFinder addonFinder) {
-        if (addonFinders.remove(addonFinder)) {
+        boolean removed;
+        synchronized (this) {
+            removed = addonFinders.remove(addonFinder);
+        }
+        if (removed) {
             changed();
         }
     }
 
-    private synchronized void changed() {
-        List<AddonInfo> candidates = addonInfoProviders.stream().map(p -> p.getAddonInfos(localeProvider.getLocale()))
-                .flatMap(Collection::stream).toList();
-        addonFinders.stream().filter(this::isFinderEnabled).forEach(f -> f.setAddonCandidates(candidates));
+    private void changed() {
+        synchronized (this) {
+            List<AddonInfo> candidates = addonInfoProviders.stream()
+                    .map(p -> p.getAddonInfos(localeProvider.getLocale())).flatMap(Collection::stream).toList();
+            addonFinders.stream().filter(this::isFinderEnabled).forEach(f -> f.setAddonCandidates(candidates));
+        }
     }
 
     @Deactivate

--- a/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
+++ b/bundles/org.openhab.core.config.discovery.addon/src/main/java/org/openhab/core/config/discovery/addon/AddonSuggestionService.java
@@ -73,7 +73,6 @@ public class AddonSuggestionService implements AutoCloseable {
     private final ScheduledExecutorService scheduler;
     private final Map<String, Boolean> baseFinderConfig = new ConcurrentHashMap<>();
     private final List<Future<?>> tasks = new CopyOnWriteArrayList<>();
-    private final Object synchObject = new Object();
 
     @Activate
     public AddonSuggestionService(final @Reference ConfigurationAdmin configurationAdmin,
@@ -162,25 +161,19 @@ public class AddonSuggestionService implements AutoCloseable {
 
     @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
     public void addAddonInfoProvider(AddonInfoProvider addonInfoProvider) {
-        synchronized (synchObject) {
-            addonInfoProviders.add(addonInfoProvider);
-        }
+        addonInfoProviders.add(addonInfoProvider);
         changed();
     }
 
     public void removeAddonInfoProvider(AddonInfoProvider addonInfoProvider) {
-        boolean removed;
-        synchronized (synchObject) {
-            removed = addonInfoProviders.remove(addonInfoProvider);
-        }
-        if (removed) {
+        if (addonInfoProviders.remove(addonInfoProvider)) {
             changed();
         }
     }
 
     @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
     public void addAddonFinder(AddonFinder addonFinder) {
-        synchronized (synchObject) {
+        synchronized (addonFinders) {
             addonFinders.add(addonFinder);
         }
         changed();
@@ -188,7 +181,7 @@ public class AddonSuggestionService implements AutoCloseable {
 
     public void removeAddonFinder(AddonFinder addonFinder) {
         boolean removed;
-        synchronized (synchObject) {
+        synchronized (addonFinders) {
             removed = addonFinders.remove(addonFinder);
         }
         if (removed) {
@@ -197,9 +190,9 @@ public class AddonSuggestionService implements AutoCloseable {
     }
 
     private void changed() {
-        synchronized (synchObject) {
-            List<AddonInfo> candidates = addonInfoProviders.stream()
-                    .map(p -> p.getAddonInfos(localeProvider.getLocale())).flatMap(Collection::stream).toList();
+        List<AddonInfo> candidates = addonInfoProviders.stream().map(p -> p.getAddonInfos(localeProvider.getLocale()))
+                .flatMap(Collection::stream).toList();
+        synchronized (addonFinders) {
             addonFinders.stream().filter(this::isFinderEnabled).forEach(f -> f.setAddonCandidates(candidates));
         }
     }
@@ -207,14 +200,14 @@ public class AddonSuggestionService implements AutoCloseable {
     @Deactivate
     @Override
     public void close() throws Exception {
-        synchronized (synchObject) {
+        synchronized (addonFinders) {
             addonFinders.clear();
-            addonInfoProviders.clear();
         }
+        addonInfoProviders.clear();
     }
 
     public Set<AddonInfo> getSuggestedAddons(@Nullable Locale locale) {
-        synchronized (synchObject) {
+        synchronized (addonFinders) {
             return addonFinders.stream().filter(this::isFinderEnabled).map(f -> f.getSuggestedAddons())
                     .flatMap(Collection::stream).collect(Collectors.toSet());
         }

